### PR TITLE
perf: draw portrait with canvas

### LINKED ISSUE #14: [Performance] Reduce memory usage when rendering Portraits
The way many sprites are rendered in Athena Crisis will render most of the image, and then cut off everything around it using CSS. This is naive, but works great for most use cases with small images that change often. However, character portraits are about 320 kb in size, and there are 11 variants of them. Rendering many characters on memory-limited devices is not good. This issue is about exploring alternative ways to render character portraits, possibly by slicing them up and rendering them with canvas and [`drawImage`](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/drawImage). There are some examples in the Athena Crisis codebase as well, see [`renderTile`](https://github.com/nkzw-tech/athena-crisis/blob/main/hera/render/renderTile.tsx).

### Steps
* Find the [`Portrait`](https://github.com/nkzw-tech/athena-crisis/blob/main/hera/character/Portrait.tsx) component.
* Implement a new rendering strategy that avoids rendering the same large image each time a portrait is used.
* Ensure that you retain all functionality such as animating the portraits. Note that not all portraits are animated at this time, so depending on which one you are trying, it may "animate" but just swap to the same image. Try it with the Pioneer, Infantry or Sniper units.

Check out the [UI Components](https://athenacrisis.com/open-source/ui/game-components#portraits) section for examples of portrait rendering.

### Funding

- We're using [Polar.sh](https://polar.sh/nkzw-tech) to distribute funds.
- You receive the reward once the issue is completed & confirmed by [Nakazawa Tech](https://nkzw.tech).

<a href="https://polar.sh/nkzw-tech/athena-crisis/issues/14">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/nkzw-tech/athena-crisis/issues/14/pledge.svg?darkmode=1">
  Fund with Polar
</picture>
</a>

### PR DESCRIPTION
**Pull request by puroong**
## Description

Resolves #14 

- store all portrait images in imageMap
  - this is necessary in order to load images with `spriteImage`
- draw portrait with canvas
- animate portrait with `setInterval`

## Memory Usage
- 1000 Portraits

### with img tag (as is)

image

### with canvas

Screenshot 2024-05-26 at 05 04 16

### CODE REVIEW

File: hera/character/Portrait.tsx [hunk 1]
```diff
@@ -30,43 +29,64 @@ export default memo(function Portrait({
...
+        PortraitWidth,
+        PortraitHeight,
+      );
+
+      curKeyFrameY = (curKeyFrameY + 1) % keyFrameYs.length;
+    }
+
+    if (animate) {
+      setInterval(animate, 1000 / keyFrameYs.length);
+    }
```
Comment (cpojer) [inline hera/character/Portrait.tsx:73]: This needs to return () => clearInterval(interval) so that it gets stopped properly.

Comment (cpojer) [reply hera/character/Portrait.tsx:73]: Also the effect needs to support the `paused` flag and ensure the animation is not started.

File: hera/character/Portrait.tsx [hunk 2]
```diff
@@ -30,43 +29,64 @@ export default memo(function Portrait({
...
+  const ref = useRef<HTMLCanvasElement>(null);
   const hasPortraits = useSprites('portraits');
-  const sprite = hasPortraits
-    ? spriteURL('Portraits', encodeDynamicPlayerID(player))
-    : '';
 
   const { position } = unit.sprite.portrait;
-  const y = -(position.y + (variant || 0)) * PortraitHeight + 'px';
-  const alternateY = -(position.y + (variant || 0) + 6) * PortraitHeight + 'px';
+  const keyFrameYs = [
```
Comment (cpojer) [inline hera/character/Portrait.tsx:36]: Maybe we can call this `positions`?

File: hera/character/Portrait.tsx [hunk 3]
```diff
@@ -30,43 +29,64 @@ export default memo(function Portrait({
...
+    (position.y + (variant || 0)) * PortraitHeight,
+    (position.y + (variant || 0) + 6) * PortraitHeight,
+  ];
+  let curKeyFrameY = 0;
```
Comment (cpojer) [inline hera/character/Portrait.tsx:40]: Let's call this `currentPosition`, and move it inside the `useLayoutEffect` call. We can't mutate variables created in render like that.

File: hera/character/Portrait.tsx [hunk 4]
```diff
@@ -30,43 +29,64 @@ export default memo(function Portrait({
...
-      )}
+      <canvas ref={ref}></canvas>
     </div>
   );
 });
 
-const vars = new CSSVariables<'x' | 'y' | 'alternate-y'>('p');
-
 const portraitStyle = css`
   contain: content;
```
Comment (cpojer) [inline hera/character/Portrait.tsx:91]: I think we don't need this line since we aren't overflowing the element anymore.

**Comment by cpojer** [comment]
@puroong wow, thanks for the contribution! This is very nice, and will make Athena Crisis work much better on low-end devices. I left a few comments to make sure we are cleaning things up.

File: hera/character/Portrait.tsx (hunk 2 above)
Comment (puroong) [reply hera/character/Portrait.tsx]: [refactor: rename variables](https://github.com/nkzw-tech/athena-crisis/pull/32/commits/9331080cfe4ad8ea0b79525f39d727743eb70cfe)

File: hera/character/Portrait.tsx (hunk 1 above)
Comment (puroong) [reply hera/character/Portrait.tsx]: - returning () => clearInterval(interval)
[feat: cleanup useLayoutEffect](https://github.com/nkzw-tech/athena-crisis/pull/32/commits/a3f0f3b271008619c35d16434845c57a9db563ca)

- supporting `paused` flag
[feat: support paused flag](https://github.com/nkzw-tech/athena-crisis/pull/32/commits/ef766b5dca0a6497adc388f361d13fc69107c173)
[fix: render image when animate, paused are not given](https://github.com/nkzw-tech/athena-crisis/pull/32/commits/8b698e192170dfdb472c572484c28b2ef1ea70d6)

File: hera/character/Portrait.tsx (hunk 4 above)
Comment (puroong) [reply hera/character/Portrait.tsx]: [feat: update portraitStyle](https://github.com/nkzw-tech/athena-crisis/pull/32/commits/e2e993f81cf79752bd117825d9730997e7b51321)

File: hera/character/Portrait.tsx (hunk 3 above)
Comment (puroong) [reply hera/character/Portrait.tsx]: [refactor: rename variables](https://github.com/nkzw-tech/athena-crisis/pull/32/commits/9331080cfe4ad8ea0b79525f39d727743eb70cfe)

**Comment by puroong** [comment]
@cpojer 
Thanks for the opportunity! I addressed the feedbacks and its ready for review 😄

**Comment by puroong** [comment]
My bad... I should have run the tests in local..
fixed lint errors in this commit!

[style: lint](https://github.com/nkzw-tech/athena-crisis/pull/32/commits/5183c968c46088f36b5eaf3ffee60eb2290cd6b8)

**Comment by cpojer** [comment]
Thank you for your contribution. The funds were distributed to you.

**Pull request merged**